File: angularapp/src/app/shared/services/bid-step.service.ts
import { Injectable } from '@angular/core';
import { BidStep } from '../contracts/requests/bid';

@Injectable({
  providedIn: 'root',
})
export class BidStepService {
  private bidSteps: BidStep[] = [
    new BidStep(0.01, 0.99, 0.05),
    new BidStep(1.0, 4.99, 0.25),
    new BidStep(5.0, 24.99, 0.5),
    new BidStep(25.0, 99.99, 1.0),
    new BidStep(100.0, 249.99, 2.5),
    new BidStep(250.0, 499.99, 5.0),
    new BidStep(500.0, 999.99, 10.0),
    new BidStep(1000.0, 2499.99, 25.0),
    new BidStep(2500.0, 4999.99, 50.0),
    new BidStep(5000.0, Number.MAX_VALUE, 100.0),
  ];

  constructor() {}

  getBidStep(currentPrice: number): number {
    const range = this.bidSteps.find(
      (b) => currentPrice >= b.minPrice && currentPrice <= b.maxPrice
    );
    if (!range) {
      throw new Error(
        `No bid increment found for the current price of ${currentPrice}`
      );
    }
    return range.step;
  }
}

File: Contracts/Requests/Bid/BidStepService.cs
namespace contracts.Requests.Bid
{
    public static class BidStepService
    {
        private static readonly List<BidStep> BidSteps;

        static BidStepService()
        {
            BidSteps = new List<BidStep>
            {
                new(0.01m, 0.99m, 0.05m),
                new(1.00m, 4.99m, 0.25m),
                new(5.00m, 24.99m, 0.50m),
                new(25.00m, 99.99m, 1.00m),
                new(100.00m, 249.99m, 2.50m),
                new(250.00m, 499.99m, 5.00m),
                new(500.00m, 999.99m, 10.00m),
                new(1000.00m, 2499.99m, 25.00m),
                new(2500.00m, 4999.99m, 50.00m),
                new(5000.00m, decimal.MaxValue, 100.00m)
            };
        }

        public static bool IsBidAmountValid(decimal currentPrice, decimal bidAmount, bool addBidStep)
        {
            var bidStep = GetBidStep(currentPrice);

            if (addBidStep)
            {
                return bidAmount >= currentPrice + bidStep;
            }

            return bidAmount >= currentPrice;
        }

        public static decimal GetBidStep(decimal currentPrice)
        {
            try
            {
                var range = BidSteps.First(b => currentPrice >= b.MinPrice && currentPrice <= b.MaxPrice);
                return range.Step;
            }
            catch (InvalidOperationException)
            {
                throw new InvalidOperationException($"No bid step found for the current price of {currentPrice:C}");
            }
        }
    }
}
